File: tools/src/translate/scoring.ts
/**
 * Humanize a `secondary-card` scoring `award` into plain English.
 *
 * Output is **ASCII-only** with a fixed clause order, pinned byte-for-byte
 * across the TS and Rust ports by the `conformance/scoring-translation` corpus.
 * The community `text` summary and the `actions` list are verbatim data, not
 * translation, so they are not produced here — only the structured `awards`.
 */

import type { SecondaryCard } from "../generated.js";
import { describeCondition, dekebab, type Condition } from "./condition.js";

/** When a VP award is evaluated (the `trigger` block on an award). */
export interface ScoringTrigger {
  timing?: "start-of-turn" | "end-of-turn" | "start-of-phase" | "end-of-phase" | "end-of-battle";
  phase?: "command" | "movement" | "shooting" | "charge" | "fight";
  player_turn?: "your-turn" | "opponent-turn" | "either";
  battle_round?: { min?: number; max?: number };
}

/** The scoring approach a card is played under (cards that print both). */
export type ScoringMode = "fixed" | "tactical";

/** One VP-award block on a scoring card. */
export interface ScoringAward {
  trigger?: ScoringTrigger;
  when?: Condition;
  vp?: number;
  vp_per?: number;
  per?: string;
  per_max?: number;
  /** Per-game VP ceiling for this award (the card's "UP TO N VP"). */
  vp_max?: number;
  cumulative?: boolean;
  exclusive_group?: string;
  /** Which scoring track this award belongs to, on cards that print both. */
  mode?: ScoringMode;
}

function capitalize(s: string): string {
  return s.length === 0 ? s : s[0].toUpperCase() + s.slice(1);
}

/** "End of your Command phase (round 2+)" and friends. */
export function describeTrigger(t: ScoringTrigger): string {
  const turn =
    t.player_turn === "opponent-turn"
      ? "the opponent's"
      : t.player_turn === "either"
        ? "any"
        : "your";

  let base: string;
  switch (t.timing) {
    case "start-of-turn":
      base = `Start of ${turn} turn`;
      break;
    case "end-of-turn":
      base = `End of ${turn} turn`;
      break;
    case "start-of-phase":
      base = `Start of ${turn} ${capitalize(t.phase ?? "")} phase`;
      break;
    case "end-of-phase":
      base = `End of ${turn} ${capitalize(t.phase ?? "")} phase`;
      break;
    case "end-of-battle":
      base = "End of the battle";
      break;
    default:
      base = t.phase ? `During ${turn} ${capitalize(t.phase)} phase` : "Any time";
  }

  const br = t.battle_round;
  if (br) {
    const { min, max } = br;
    if (min != null && max != null) {
      base += min === max ? ` (round ${min})` : ` (rounds ${min}-${max})`;
    } else if (min != null) {
      base += ` (round ${min}+)`;
    } else if (max != null) {
      base += ` (rounds 1-${max})`;
    }
  }
  return base;
}

/** "End of your Command phase (round 2+): 3 VP per controlled objective when ..." */
export function describeAward(a: ScoringAward): string {
  const trigger = a.trigger ? describeTrigger(a.trigger) : "Any time";

  let amount: string;
  if (a.vp != null) {
    amount = `${a.vp} VP`;
  } else if (a.vp_per != null) {
    amount = `${a.vp_per} VP per ${a.per ? dekebab(a.per) : "instance"}`;
    if (a.per_max != null) amount += ` (max ${a.per_max})`;
  } else {
    amount = "no VP";
  }

  const prefix = a.cumulative ? "+ " : "";
  const when = a.when ? ` when ${describeCondition(a.when)}` : "";
  const tier = a.exclusive_group ? " [highest tier]" : "";
  return `${prefix}${trigger}: ${amount}${when}${tier}`;
}

/** Humanize every award on a card, in array order (the order is load-bearing). */
export function describeScoringCard(card: SecondaryCard): string[] {
  const awards = (card.awards ?? []) as unknown as ScoringAward[];
  return awards.map(describeAward);
}

File: python/src/wh40kdc/translate/scoring.py
"""Humanize a ``secondary-card`` scoring ``award`` into plain English.

Output is **ASCII-only** with a fixed clause order, pinned byte-for-byte
across the TS, Rust, and Python ports by the
``conformance/scoring-translation`` corpus. The community ``text`` summary
and the ``actions`` list are verbatim data, not translation, so they are not
produced here — only the structured ``awards``.

Python mirror of ``tools/src/translate/scoring.ts``.
"""

from __future__ import annotations

from typing import Any

from wh40kdc.translate.condition import dekebab, describe_condition


def _capitalize(s: str) -> str:
    return s if s == "" else s[0].upper() + s[1:]


def describe_trigger(t: dict[str, Any]) -> str:
    """"End of your Command phase (round 2+)" and friends."""
    player_turn = t.get("player_turn")
    if player_turn == "opponent-turn":
        turn = "the opponent's"
    elif player_turn == "either":
        turn = "any"
    else:
        turn = "your"

    timing = t.get("timing")
    phase = t.get("phase")
    if timing == "start-of-turn":
        base = f"Start of {turn} turn"
    elif timing == "end-of-turn":
        base = f"End of {turn} turn"
    elif timing == "start-of-phase":
        base = f"Start of {turn} {_capitalize(phase or '')} phase"
    elif timing == "end-of-phase":
        base = f"End of {turn} {_capitalize(phase or '')} phase"
    elif timing == "end-of-battle":
        base = "End of the battle"
    else:
        base = f"During {turn} {_capitalize(phase)} phase" if phase else "Any time"

    br = t.get("battle_round")
    if br:
        min_ = br.get("min")
        max_ = br.get("max")
        if min_ is not None and max_ is not None:
            base += f" (round {min_})" if min_ == max_ else f" (rounds {min_}-{max_})"
        elif min_ is not None:
            base += f" (round {min_}+)"
        elif max_ is not None:
            base += f" (rounds 1-{max_})"
    return base


def describe_award(a: dict[str, Any]) -> str:
    """"End of your Command phase (round 2+): 3 VP per controlled objective when ..." """
    trigger = describe_trigger(a["trigger"]) if a.get("trigger") else "Any time"

    if a.get("vp") is not None:
        amount = f"{a['vp']} VP"
    elif a.get("vp_per") is not None:
        per = dekebab(a["per"]) if a.get("per") else "instance"
        amount = f"{a['vp_per']} VP per {per}"
        if a.get("per_max") is not None:
            amount += f" (max {a['per_max']})"
    else:
        amount = "no VP"

    prefix = "+ " if a.get("cumulative") else ""
    when = f" when {describe_condition(a['when'])}" if a.get("when") else ""
    tier = " [highest tier]" if a.get("exclusive_group") else ""
    return f"{prefix}{trigger}: {amount}{when}{tier}"


def describe_scoring_card(card: dict[str, Any]) -> list[str]:
    """Humanize every award on a card, in array order (order is load-bearing)."""
    return [describe_award(a) for a in card.get("awards") or []]
